design-proposals/klip-35-dynamic-processing-log-levels.md

```diff
@@ -0,0 +1,210 @@
+- Kafka-based stateful implementation that distributes level settings to the cluster and stores them so they are recovered on a restart.
+
+## Out-Of-Scope
+
+- Thoroughly populating the query engine with log messages to address the problems we’ve already seen. This is super important, but out of scope for this particular KLIP.
+
+## Value/Return
+
+A step toward improving UX when users observe unexpected behavior from running queries.
+
+## Public APIs
+
+### REST API
+
+### PUT /processing_logger/config
+
+Sets the root processing log configuration (currently just supports setting the log level).  This will set the level for all loggers that do not already have a level set.
+
+Request Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+````
+
+Response Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level from the request”}
+    }
+}
+```
+
+### GET /processing_logger/config
+
+Gets the current root processing log level.
+
+Response Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level of the root logger”}
+    }
+}
+```
+
+### DELETE /processing_logger/config
+
+Deletes customized config for the root logger (reverts to default config as specified in the log4j config file).
+
+### PUT /processing_logger/[processing logger name]/config
+
+Sets the processing log configuration for the logger with the given processing logger name (currently just supports setting the log level). This will set the level for all child loggers (loggers that share a common name prefix) that do not already have a level set using a longer prefix.
+
+Request Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+```
+
+Response Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level from the request”}
+    }
+}
+```
+
+### GET /processing_logger/[processing logger name]/level
+
+Gets the processing log configuration for the logger with the given processing logger name.
+
+Response Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level of the specified logger”}
+    }
+}
+```
+
+### DELETE /processing_logger/[processing logger name]/level
+
+Deletes customized config for the logger with the given processing logger name.
+
+### GET /processing_logger/[processing logger name?]
+
+Lists all processing loggers and configs under the logger name if specified. If not specified, lists all loggers and configs.
+
+Response Fields:
+```
+{
+    “properties”: {
+        “loggers”: {
+            “type”: “array”,
+            “items”: {
+                “type" : “object”,
+                "properties" : “{
+                     “context" : {“type”: “string”, “description”: “the context of this logger”}
```

Comment:
Double-checking: I assume logger context and logger name are the same? And also "query context" below?

Reply:
yeah - I'll just change all the X context instances to logger name. Mixed up my nomenclature.